File: app/models/academy_photo.py
"""Models para OctoPhotos — feed de fotos por academia (feature premium)."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models.base import UUIDMixin
from app.models.soft_delete import SoftDeleteMixin


class AcademyPhoto(Base, UUIDMixin, SoftDeleteMixin):
    """Post de foto no feed da academia."""

    __tablename__ = "academy_photos"
    __table_args__ = (
        # Índice parcial cobre o feed paginado (academy_id + created_at DESC, só posts ativos)
        Index("ix_academy_photos_feed", "academy_id", "created_at", postgresql_where="deleted_at IS NULL"),
        Index("ix_academy_photos_status", "status", postgresql_where="deleted_at IS NULL"),
    )

    academy_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("academies.id", ondelete="CASCADE"),
        nullable=False,
    )
    author_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    image_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    thumbnail_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    raw_file_path: Mapped[str | None] = mapped_column(Text, nullable=True)
    caption: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="processing")
    likes_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    comments_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_system_post: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    system_post_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    system_post_ref_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)

    author: Mapped[User] = relationship("User", foreign_keys=[author_id], lazy="selectin")
    likes: Mapped[list[AcademyPhotoLike]] = relationship(
        "AcademyPhotoLike", back_populates="photo", lazy="raise", passive_deletes=True
    )


class AcademyPhotoComment(Base, UUIDMixin):
    """Comentário em um post de foto."""

    __tablename__ = "academy_photo_comments"
    __table_args__ = (
        Index("ix_photo_comments_photo", "photo_id", "created_at", postgresql_where="deleted_at IS NULL"),
    )

    photo_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("academy_photos.id", ondelete="CASCADE"),
        nullable=False,
    )
    author_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    body: Mapped[str] = mapped_column(Text, nullable=False)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    author: Mapped[User] = relationship("User", foreign_keys=[author_id], lazy="selectin")


class AcademyPhotoLike(Base):
    """Curtida de um usuário em um post de foto (PK composta, sem UUID próprio)."""

    __tablename__ = "academy_photo_likes"

    photo_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("academy_photos.id", ondelete="CASCADE"),
        primary_key=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        primary_key=True,
        index=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    photo: Mapped[AcademyPhoto] = relationship("AcademyPhoto", back_populates="likes")


class AcademyPhotoRestriction(Base, UUIDMixin):
    """Restrição de postagem de um aluno na academia (gerente/admin podem restringir temporária ou permanentemente)."""

    __tablename__ = "academy_photo_restrictions"

    academy_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("academies.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    restricted_by: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=False,
    )
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    restricted_user: Mapped[User] = relationship("User", foreign_keys=[user_id], lazy="selectin")
    restricted_by_user: Mapped[User] = relationship("User", foreign_keys=[restricted_by], lazy="selectin")

File: viewer/lib/models/academy_photo.dart
class PhotoAuthor {
  final String id;
  final String? name;
  final String? avatarUrl;

  const PhotoAuthor({required this.id, this.name, this.avatarUrl});

  factory PhotoAuthor.fromJson(Map<String, dynamic> json) {
    return PhotoAuthor(
      id: json['id'] as String,
      name: json['name'] as String?,
      avatarUrl: json['avatar_url'] as String?,
    );
  }
}

class PhotoComment {
  final String id;
  final String photoId;
  final PhotoAuthor author;
  final String body;
  final DateTime createdAt;

  const PhotoComment({
    required this.id,
    required this.photoId,
    required this.author,
    required this.body,
    required this.createdAt,
  });

  factory PhotoComment.fromJson(Map<String, dynamic> json) {
    return PhotoComment(
      id: json['id'] as String,
      photoId: json['photo_id'] as String,
      author: PhotoAuthor.fromJson(json['author'] as Map<String, dynamic>),
      body: json['body'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class AcademyPhoto {
  final String id;
  final String academyId;
  final PhotoAuthor author;
  final String? imageUrl;
  final String? thumbnailUrl;
  final String? caption;
  final String status;
  final int likesCount;
  final int commentsCount;
  final bool likedByMe;
  final bool isSystemPost;
  final String? systemPostType;
  final String? systemPostRefId;
  final DateTime createdAt;

  const AcademyPhoto({
    required this.id,
    required this.academyId,
    required this.author,
    this.imageUrl,
    this.thumbnailUrl,
    this.caption,
    required this.status,
    required this.likesCount,
    this.commentsCount = 0,
    required this.likedByMe,
    required this.isSystemPost,
    this.systemPostType,
    this.systemPostRefId,
    required this.createdAt,
  });

  bool get isReady => status == 'ready';

  factory AcademyPhoto.fromJson(Map<String, dynamic> json) {
    return AcademyPhoto(
      id: json['id'] as String,
      academyId: json['academy_id'] as String,
      author: PhotoAuthor.fromJson(json['author'] as Map<String, dynamic>),
      imageUrl: json['image_url'] as String?,
      thumbnailUrl: json['thumbnail_url'] as String?,
      caption: json['caption'] as String?,
      status: json['status'] as String? ?? 'processing',
      likesCount: json['likes_count'] as int? ?? 0,
      commentsCount: json['comments_count'] as int? ?? 0,
      likedByMe: json['liked_by_me'] as bool? ?? false,
      isSystemPost: json['is_system_post'] as bool? ?? false,
      systemPostType: json['system_post_type'] as String?,
      systemPostRefId: json['system_post_ref_id'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  AcademyPhoto copyWith({
    String? id,
    String? academyId,
    PhotoAuthor? author,
    String? imageUrl,
    String? thumbnailUrl,
    String? caption,
    String? status,
    int? likesCount,
    int? commentsCount,
    bool? likedByMe,
    bool? isSystemPost,
    String? systemPostType,
    String? systemPostRefId,
    DateTime? createdAt,
  }) {
    return AcademyPhoto(
      id: id ?? this.id,
      academyId: academyId ?? this.academyId,
      author: author ?? this.author,
      imageUrl: imageUrl ?? this.imageUrl,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      caption: caption ?? this.caption,
      status: status ?? this.status,
      likesCount: likesCount ?? this.likesCount,
      commentsCount: commentsCount ?? this.commentsCount,
      likedByMe: likedByMe ?? this.likedByMe,
      isSystemPost: isSystemPost ?? this.isSystemPost,
      systemPostType: systemPostType ?? this.systemPostType,
      systemPostRefId: systemPostRefId ?? this.systemPostRefId,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

class PhotoFeedPage {
  final List<AcademyPhoto> items;
  final String? nextCursor;

  const PhotoFeedPage({required this.items, this.nextCursor});

  factory PhotoFeedPage.fromJson(Map<String, dynamic> json) {
    final rawItems = json['items'] as List<dynamic>? ?? [];
    return PhotoFeedPage(
      items: rawItems
          .map((e) => AcademyPhoto.fromJson(e as Map<String, dynamic>))
          .toList(),
      nextCursor: json['next_cursor'] as String?,
    );
  }
}

class PhotoRestriction {
  final String id;
  final String academyId;
  final String userId;
  final String? userName;
  final String? reason;
  final DateTime? expiresAt;
  final bool active;
  final DateTime createdAt;

  const PhotoRestriction({
    required this.id,
    required this.academyId,
    required this.userId,
    this.userName,
    this.reason,
    this.expiresAt,
    required this.active,
    required this.createdAt,
  });

  factory PhotoRestriction.fromJson(Map<String, dynamic> json) {
    return PhotoRestriction(
      id: json['id'] as String,
      academyId: json['academy_id'] as String,
      userId: json['user_id'] as String,
      userName: json['user_name'] as String?,
      reason: json['reason'] as String?,
      expiresAt: json['expires_at'] != null
          ? DateTime.parse(json['expires_at'] as String)
          : null,
      active: json['active'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
